### ISSUE TITLE
Stack Overflow in BindPatternSwitchLabelForInference

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#16559]
**Version Used**: VS2015 RC3

**Steps to Reproduce**:

Attempt to compile this code:

```cs
public class Program5815
{
    public static void Main(string[] args)
    {
        switch ((int)M())
        {
            case var x3:
            case true ? x3 : 4:
                break;
        }
    }
    private static object M() => null;
}
```

**Expected Behavior**: Compiler error about case label not being a constant

**Actual Behavior**: Compiler stack overflow

See also https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=369796

**Comment by jcouv** [comment]
Tracked by VSO https://devdiv.visualstudio.com/DevDiv/_workitems?id=369796&_a=edit

**Comment by jcouv** [comment]
Fixed by  #16569

**Comment by gafter** [comment]
Not fixed in #16569. That only adds a skipped test.
